plugins/woocommerce-blocks/assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,140 +1,319 @@
+
+	/**
+	 * Private constructor to enforce singleton pattern.
+	 * Initializes subscriptions for template changes.
+	 */
+	private constructor() {
+		this.initializeSubscriptions();
 	}
 
-	const editSiteStore = select( 'core/edit-site' );
+	/**
+	 * Gets the singleton instance of the BlockRegistrationManager.
+	 * Creates the instance if it doesn't exist.
+	 *
+	 * @return {BlockRegistrationManager} The singleton instance
+	 */
+	public static getInstance(): BlockRegistrationManager {
+		if ( ! BlockRegistrationManager.instance ) {
+			BlockRegistrationManager.instance = new BlockRegistrationManager();
+		}
+		return BlockRegistrationManager.instance;
+	}
 
-	if ( editSiteStore ) {
-		subscribe( () => {
-			const previousTemplateId = currentTemplateId;
+	/**
+	 * Parses a template ID from various possible formats.
+	 * Handles both string and number inputs due to Gutenberg changes.
+	 *
+	 * @param {string | number | undefined} templateId - The template ID to parse
+	 * @return {string | undefined} The parsed template ID
+	 */
+	private parseTemplateId(
+		templateId: string | number | undefined
+	): string | undefined {
+		const parsedTemplateId = isNumber( templateId )
+			? undefined
+			: templateId;
+		return parsedTemplateId?.split( '//' )[ 1 ];
+	}
 
-			// With GB 16.3.0 the return type can be a number: https://github.com/WordPress/gutenberg/issues/53230
-			currentTemplateId = parseTemplateId(
-				editSiteStore?.getEditedPostId< string | number | undefined >()
-			);
-			const hasChangedTemplate = previousTemplateId !== currentTemplateId;
+	/**
+	 * Initializes subscriptions for template changes and block registration.
+	 * Sets up listeners for both the site editor and post editor contexts.
+	 */
+	private initializeSubscriptions(): void {
+		if ( this.initialized ) {
+			return;
+		}
 
-			if ( ! hasChangedTemplate || ! blockName ) {
+		// Site Editor Subscription
+		const siteEditorUnsubscribe = subscribe( () => {
+			const editSiteStore = select( 'core/edit-site' );
+			if ( ! editSiteStore ) {
 				return;
 			}
 
-			let isRegistered = isBlockRegistered( blockName );
-
-			/**
-			 * We need to unregister the block each time the user visits or leaves the Single Product template.
-			 *
-			 * The Single Product template is the only template where the `ancestor` property is not needed because it provides the context
-			 * for the product blocks. We need to unregister and re-register the block to remove or add the `ancestor` property depending on which
-			 * location (template, post, page, etc.) the user is in.
-			 *
-			 */
-			if (
-				isRegistered &&
-				( currentTemplateId?.includes( 'single-product' ) ||
-					previousTemplateId?.includes( 'single-product' ) )
-			) {
-				if ( isVariationBlock && variationName ) {
-					unregisterBlockVariation( blockName, variationName );
-				} else {
-					unregisterBlockType( blockName );
+			// Only run this once for site editor
+			siteEditorUnsubscribe();
```

Comment:
The `siteEditorUnsubscribe()` is called right after checking for the edit site store because we only want to set up the template change listener once. Otherwise the below code would resubscribe on every store update.

```
        // This subscription would run EVERY time the store updates
        subscribe(() => {
            const previousTemplateId = this.currentTemplateId;
            this.currentTemplateId = this.parseTemplateId(
                editSiteStore?.getEditedPostId<string | number | undefined>()
            );

            if (previousTemplateId !== this.currentTemplateId) {
                this.handleTemplateChange(previousTemplateId);
            }
        }, 'core/edit-site');
```

That said, it should just be enough to check for the stores existence like

```
 const editSiteStore = select('core/edit-site');
 if (editSiteStore) {
 ...
 }
 ```
 
 So I'll make that update.

Reply:
Updated in https://github.com/woocommerce/woocommerce/pull/53788/commits/dabdb45b208dd7c1dc4874c1816282bba51bccaf